designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
+  greyed-out to indicate that it is stale. It is not minimized.
+- When a bucket or object is created or deleted, only that node is added or
+  removed from the tree, _without disturbing the current layout of the tree_.
+
+
+![](20200210-design-vscode-s3-tree.png)
+
+### Create and edit S3 buckets
+
+- Root node context menu (right-click):
+  - Create bucket
+- Bucket node context menu (right-click):
+  - Delete bucket
+  - New folder...
+  - Upload...
+  - Copy name
+  - Copy ARN
+
+### Create and edit S3 objects
+
+- S3 object node context menu (right-click):
+  - Delete object
+  - Copy name
+  - Copy ARN
+  - Copy URL
+  - Upload...
+  - Edit
+    - Downloads the object to a temp folder and opens it in a new VSCode editor tab.
+  - Download
+    - Does _not_ present a chooser.
+    - Downloads immediately to the OS default "Downloads" location (typically `~/Downloads`).
+    - Download status is given as a toaster, which contains a "Choose location..." link.
+      - User can change the download location via this link. This will be the
+        default location for future downloads.
+      - The default download location is also configurable as a global Toolkit
+        option. This is not S3-specific, it is used for all "Download"
+        experiences in the Toolkit.
+
+### Drag-drop from client to server
+
+- Drag a folder from the local system to the S3 tree:
+  - onto a bucket uploads as a new top-level folder in the bucket
+  - onto a folder uploads as a new subfolder
+- Drag a file from the local system to the S3 tree:
+  - onto a bucket uploads the file as a top-level object in the bucket
+  - onto a folder uploads the file as an object in the folder
+
+### Drag-drop server-side
+
+- User can copy objects or folders from one folder to another by drag-drop.
+- User can copy objects or folders _across regions_ by dragging from one region
+  to another.
+
+### S3 bucket properties
+
+Surface popular bucket properties such as "public accessible" and "versioning
+enabled" in a clear yet uncluttered manner.
+
+- Show icon (same as AWS web console) next to bucket name indicating whether it
+  is publicly accessible
+- Show icon or sigil next to bucket name indicating whether versioning is enabled
+
+### S3 object properties
+
+S3 objects have properties like size, last-modified-date, and previous versions
+(if versioning is enabled).
+
+- If >=1 previous versions are available for an object, show the count of
+  previous versions next to the object.
+- If >=1 previous versions are available for an object, context menu
+  (right-click) on the object exposes a "Show previous versions" menu item
+  (otherwise greyed-out).
+  - Previous versions are presented as children of the object.
```

Comment:
Can we take all of the same actions on previous versions that we can on the present version? If not, consider moving this to a nested context menu (we do something similar for Cloud Debug in the Jetbrains toolkit):
```
Previous Versions... > 1
                       2
                       3
                       ...
```
(that being said, I don't think this would work with the tooltip suggestion above)

Reply:
I don't know if VS Code supports nested context menus

Reply:
Nested menus in general are a higher-friction experience.

Reply:
If an object has a ton of previous versions, it may clutter the tree too much. Depending on the available actions, it might also be worth considering putting this into a webview where you can choose your older version from a dropdown (you could also provide additional metadata this way). I THINK this might be how Jetbrains does it?